src/main/java/bio/terra/cli/app/AuthenticationManager.java

```diff
@@ -0,0 +1,200 @@
+      // and can use later to look up the persisted Google (user) credentials
+      String cliGeneratedUserKey = UUID.randomUUID().toString();
+      userLoggingIn = new TerraUser(cliGeneratedUserKey);
+    }
+
+    // fetch the user credentials, prompt for login and consent if they do not already exist or are
+    // expired.
+    UserCredentials userCredentials;
+    try (InputStream inputStream =
+        AuthenticationManager.class.getClassLoader().getResourceAsStream(CLIENT_SECRET_FILENAME)) {
+
+      // if there are already credentials for this user, and they are not expired, then return them
+      // otherwise, log the user in and get their consent
+      userCredentials =
+          AuthenticationUtils.doLoginAndConsent(
+              userLoggingIn.cliGeneratedUserKey,
+              SCOPES,
+              inputStream,
+              globalContext.resolveGlobalContextDir().toFile());
+    } catch (IOException | GeneralSecurityException ex) {
+      throw new RuntimeException("Error fetching user credentials.", ex);
+    }
+    userLoggingIn.userCredentials = userCredentials;
+
+    // fetch the user information from SAM (subject id = terra user id, email = terra user name)
+    SAMUtils.populateTerraUserInfo(userLoggingIn, globalContext);
+
+    // fetch the pet SA credentials from SAM
+    HTTPUtils.HTTPResponse petSAKeySAMResponse = SAMUtils.getPetSAKey(userLoggingIn, globalContext);
+    if (petSAKeySAMResponse.statusCode != 200) {
+      throw new RuntimeException("Error fetching pet SA key from SAM");
+    }
+    try {
+      // persist it in the global context directory
+      Path jsonKeyFile =
+          FileUtils.writeStringToFile(
+              globalContext.resolvePetSAKeyDir(),
+              userLoggingIn.terraUserId,
+              petSAKeySAMResponse.responseBody);
+
+      // create a credentials object from the key
+      ServiceAccountCredentials petSACredentials =
+          AuthenticationUtils.getServiceAccountCredential(jsonKeyFile.toFile(), SCOPES);
+
+      userLoggingIn.petSACredentials = petSACredentials;
+    } catch (IOException ioEx) {
+      logger.error("Error writing pet SA key to the global context directory.", ioEx);
+    }
+
+    // update the global context with the current user
+    // note that this state is persisted to disk. it will be useful for code called in the same or a
+    // later CLI command/process
+    globalContext.addOrUpdateTerraUser(userLoggingIn);
+    globalContext.setCurrentTerraUser(userLoggingIn);
+    globalContext.writeToFile();
+  }
+
+  /** Delete all credentials associated with the current Terra user. */
+  public void logoutTerraUser() {
+    if (globalContext == null) {
+      throw new RuntimeException("Error reading global context.");
+    }
+    if (globalContext.getCurrentTerraUser() == null) {
+      logger.debug("There is no current Terra user defined in the global context.");
+      return;
+    }
+
+    try (InputStream inputStream =
```

Comment:
I had it that way originally, but it doesn't save any code because the 3 try-with-resources calls in AuthManager are calling 3 separate methods in AuthUtils. So passing a string here just moves the try-with-resources to a different class, and I slightly prefer having the client secret read in from the class path in the AuthManager rather than the utils.

I agree about finding a way to reduce boilerplate, just not sure the best approach here.

Reply:
Yeah, it does sound a little tricky and maybe not worth investment unless it becomes used more widely elsewhere too. If it's isolated to just the auth commands, maybe it doesn't cross that threshold.

But maybe a low-effort approach would be a separate method which catches the checked exceptions and returns the parsed GoogleClientSecrets, e.g.

```
private GoogleClientSecrets loadClientSecret() {
  try {
    InputStream input = {class loader stuff};
    GoogleClientSecrets clientSecrets = {parse JSON stuff};
    return clientSecrets;
  } catch (IOException | GeneralSecurityException ex) {
    throw new RuntimeException("Error loading client secret JSON", ex);
  }
}
```

Then each call site would become

```
AuthenticationUtils.deleteExistingCredential(
          globalContext.getCurrentTerraUser().cliGeneratedUserKey,
          SCOPES,
          loadClientSecret(),
          globalContext.resolveGlobalContextDir().toFile());
```

Reply:
I don't think that's really equivalent. The try-with-resources will close the stream if the AuthUtils.doLoginAndConsent (for example) throws an exception or the stream opening throws an exception. I think separating the client secret read to a different method would only close the stream in the latter case.

Reply:
Maybe I misunderstood the intent there. Do we really care about managing the client-secret JSON input stream by the time we're fetching user credentials? I would have thought these are two totally separate operations with independent exception handling.

(I do see though, that my loadClientSecret method above doesn't close the input stream on exception; turning that into a try-with-resources should work.)